### DISCUSSION TITLE
Only exposing necessary information

### DISCUSSION THREAD

**Discussion by tumbl3w33d** [kanidm/kanidm#2671] [General]
I'd argue it would make sense to, by default, not expose the running kani version (currently sent as response header). I don't see how this helps anyone (but developers or someone actively debugging) and in that case it should be something you deliberately activate. You might just as well execute `kanidmd version` in the running container. If you want to monitor what version was running at a certain point in time, I think this should be part of your (internal) metrics endpoint.

Actually it's the same question about the swagger endpoint being active by default, however, that seems less critical. Curious to hear your opinions about it.

Basic principle is that you only expose as much information as necessary to not make it easier for attackers (especially automated information gathering).

**Comment by yaleman** [comment]
The response header's currently used as a check to ensure the client isn't doing weird things against the wrong version, and yes both that and the swagger docs could be used to infer the version... 

... but so could hashing the WASM returned for the UI and a variety of other things? 🤔

**Reply by tumbl3w33d** [reply]
I see. So if I go full paranoid and drop the header in my proxy, do I risk malfunction of the client?

**Reply by yaleman** [reply]
It will just whinge about the version mismatch, that's all - https://github.com/kanidm/kanidm/blob/12900eee7edb42d69847079389cd6611a1fded71/libs/client/src/lib.rs#L592-L620

**Comment by Firstyear** [comment]
> Basic principle is that you only expose as much information as necessary to not make it easier for attackers (especially automated information gathering).

I'm going to challenge this notion. Similar to Kerckhoffs principle, knowledge of a system should not weaken the system.

Knowing the version of Kanidm does not improve the situation for an attacker. Inversely, even if they didn't know the version, there are still ways to fingerprint the version based on feature availability. And finally, you don't need to know the version to attempt exploits (if they existed) against the server.

As a result, the version does not change anything for an attacker IMO.

**Reply by tumbl3w33d** [reply]
I agree, but it's what you keep hearing from security people and written in penetration test reports here. I think it's on the level of moving a default port of an app to dodge some scripts and basically it's fine for me to follow such advice when it doesn't cost anything. But now I know you connect the client's compatibility check to it.

**Reply by Firstyear** [reply]
Then you hire better security people. :) 

And yes, we use it to check for compatibility with the client, so we can warn if there will be issues, so removing it would break that and potentially lead to other issues.
